common/web/lm-worker/src/correction/distance-modeler.ts

```diff
@@ -537,6 +535,136 @@ namespace correction {
+        /**
+         * The system time when this instance was created.
+         */
+        private start: number;
+
+        /**
+         * Marks the system time at the start of the currently-running loop, as noted
+         * by a call to the `startLoop` function.
+         */
+        private loopStart: number;
+
+        private maxExecutionTime: number;
+        private maxTrueTime: number;
+
+        private executionTime: number;
+
+        /**
+         * Used to track intervals in which potential context swaps by the OS may
+         * have occurred.  Context switches generally seem to pause threads for
+         * at least 16 ms, while we expect each loop iteration to complete
+         * within just 1 ms.  So, any possible context switch should have the
+         * longest observed change in system time.
+         *
+         * See `updateOutliers` for more details.
+         */
+        private largestIntervals: number[] = [0];
+
+        constructor(maxExecutionTime: number, maxTrueTime: number) {
+          // JS measures time by the number of milliseconds since Jan 1, 1970.
+          this.loopStart = this.start = Date.now();
+          this.maxExecutionTime = maxExecutionTime;
+          this.maxTrueTime = maxTrueTime;
+        }
+
+        startLoop() {
+          this.loopStart = Date.now();
+        }
+
+        markIteration() {
+          const now = Date.now();
+          const delta = now - this.loopStart;
+          this.executionTime += delta;
+
+          /**
+           * Update the list of the three longest system-time intervals observed
+           * for execution of a single loop iteration.
+           *
+           * Ignore any zero-ms length intervals; they'd make the logic much
+           * messier than necessary otherwise.
+           */
+          if(delta) {
+            // If the currently-observed interval is longer than the shortest of the 3
+            // previously-observed longest intervals, replace it.
+            if(this.largestIntervals.length > 2 && delta > this.largestIntervals[0]) {
+              this.largestIntervals[0] = delta;
+            } else {
+              this.largestIntervals.push(delta);
+            }
+            // Puts the list in ascending order.  Shortest of the list becomes the head,
+            // longest one the tail.
+            this.largestIntervals.sort();
+
+            // Then, determine if we need to update our outlier-based tweaks.
+            this.updateOutliers();
+          }
+        }
+
+        updateOutliers() {
+          /* Base assumption:  since each loop of the search should evaluate within ~1ms,
+           *                   notably longer execution times are probably context switches.
+           *
+           * Base assumption:  OS context switches generally last at least 16ms.  (Based on
+           *                   a window.setTimeout() usually not evaluating for at least
+           *                   that long, even if set to 1ms.)
```

Comment:
Basis:  try running this code in your browser's developer mode a few times:

```
var now = Date.now();  var timer = window.setTimeout(() => {console.log(Date.now() - now)}, 1);
```

Of course, keep in mind that this code will be triggered via user-input interrupt, so it can easily be triggered while the relevant thread would otherwise be asleep, possibly in the middle of whatever timer mechanism the OS uses to periodically context-switch.  You will see notably lower values than 16ms, but you'll likely also see some higher values.  

Example runs on Win 10 / Firefox:

![image](https://user-images.githubusercontent.com/25213402/183548606-3ed9c3d6-59a5-4943-b0f0-32904b6e1c2a.png)

If a 23ms context-switch pause occurred during a prediction search, _with no other interruptions_... that'd cut execution time from 33ms to just 10ms without the special handling added here.  If it were paused even longer, well... good look operating with literally 0ms allotted execution time.  As noted by https://github.com/keymanapp/keyman/issues/5467 / https://github.com/keymanapp/keyman/pull/5479, it can happen with some regularity.

While we don't know the exact conditions of the BrowserStack servers, devices, etc used for unit testing... I wouldn't be surprised if they have at least a few VMs set up per server; if so, there's a pretty solid chance of significant context-switching happening on those systems.

Reply:
https://developer.mozilla.org/en-US/docs/Web/API/setTimeout#reasons_for_delays_longer_than_specified may also be helpful background.

On Windows, [this](https://docs.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-gettickcount) may also be relevant:

> The resolution of the GetTickCount function is limited to the resolution of the system timer, which is typically in the range of 10 milliseconds to 16 milliseconds. The resolution of the GetTickCount function is not affected by adjustments made by the [GetSystemTimeAdjustment](https://docs.microsoft.com/en-us/windows/desktop/api/sysinfoapi/nf-sysinfoapi-getsystemtimeadjustment) function.

I've got no idea what sort of behaviour we are going to see on mobile devices for these scenarios. Sounds like some testing on low-end devices is seriously needed.